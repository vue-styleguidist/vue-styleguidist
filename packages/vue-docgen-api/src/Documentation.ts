import Map from 'ts-map'
import {
	Descriptor,
	PropDescriptor,
	MethodDescriptor,
	SlotDescriptor,
	EventDescriptor,
	ComponentDoc,
	DocBlockTags,
	BlockTag,
	Param,
	UnnamedParam,
	Tag,
	ParamTag,
	ParamType
} from 'vue-inbrowser-compiler-utils'

export {
	Descriptor,
	PropDescriptor,
	MethodDescriptor,
	SlotDescriptor,
	EventDescriptor,
	ComponentDoc,
	DocBlockTags,
	BlockTag,
	Param,
	UnnamedParam,
	Tag,
	ParamTag,
	ParamType
}

export default class Documentation {
	private propsMap: Map<string, PropDescriptor>
	private methodsMap: Map<string, MethodDescriptor>
	private slotsMap: Map<string, SlotDescriptor>
	private eventsMap: Map<string, any>
	private dataMap: Map<string, any>
	private docsBlocks: string[] | undefined
	private originExtendsMixin: Descriptor
	public readonly componentFullfilePath: string

	public constructor(fullFilePath: string) {
		this.componentFullfilePath = fullFilePath
		this.propsMap = new Map()
		this.methodsMap = new Map()
		this.slotsMap = new Map()
		this.eventsMap = new Map()
		this.originExtendsMixin = {}

		this.dataMap = new Map()
	}

	public setOrigin(origin: Descriptor) {
		this.originExtendsMixin = origin.extends ? { extends: origin.extends } : {}

		this.originExtendsMixin = origin.mixin ? { mixin: origin.mixin } : {}
	}

	public setDocsBlocks(docsBlocks: string[]) {
		this.docsBlocks = docsBlocks
	}

	public set(key: string, value: any) {
		this.dataMap.set(key, value)
	}

	public get(key: string): any {
		return this.dataMap.get(key)
	}

	public getPropDescriptor(propName: string): PropDescriptor {
		const vModelDescriptor = this.propsMap.get('v-model')
		return vModelDescriptor && vModelDescriptor.name === propName
			? vModelDescriptor
			: this.getDescriptor(propName, this.propsMap, () => ({
					name: propName
			  }))
	}

	public getMethodDescriptor(methodName: string): MethodDescriptor {
		return this.getDescriptor(methodName, this.methodsMap, () => ({
			name: methodName
		}))
	}

	public getEventDescriptor(eventName: string): EventDescriptor {
		return this.getDescriptor(eventName, this.eventsMap, () => ({
			name: eventName
		}))
	}

	public getSlotDescriptor(slotName: string): SlotDescriptor {
		return this.getDescriptor(slotName, this.slotsMap, () => ({
			name: slotName
		}))
	}

	public toObject(): ComponentDoc {
		const props = this.getObjectFromDescriptor(this.propsMap)
		const methods = this.getObjectFromDescriptor(this.methodsMap)
		const events = this.getObjectFromDescriptor(this.eventsMap)
		const slots = this.getObjectFromDescriptor(this.slotsMap)

		const obj: { [key: string]: any } = {}
		this.dataMap.forEach((value, key) => {
			if (key) {
				obj[key] = value
			}
		})

		if (this.docsBlocks) {
			obj.docsBlocks = this.docsBlocks
		}

		return {
			...obj,
			// initialize non null params
			description: obj.description || '',
			tags: obj.tags || {},

			// set all the static properties
			exportName: obj.exportName,
			displayName: obj.displayName,
			props,
			events,
			methods,
			slots
		}
	}

	private getDescriptor<T>(name: string, map: Map<string, T>, init: () => T): T {
		let descriptor = map.get(name)
		if (!descriptor) {
			descriptor = init()
			descriptor = { ...descriptor, ...this.originExtendsMixin }
			map.set(name, descriptor)
		}
		return descriptor
	}

	private getObjectFromDescriptor<T>(map: Map<string, T>): T[] | undefined {
		if (map.size > 0) {
			const obj: T[] = []
			map.forEach((descriptor, name) => {
				if (name && descriptor) {
					obj.push(descriptor)
				}
			})
			return obj
		} else {
			return undefined
		}
	}
}
