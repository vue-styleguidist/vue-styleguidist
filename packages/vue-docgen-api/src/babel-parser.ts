import { parse, ParserOptions } from '@babel/parser'
import * as bt from '@babel/types'

const babelParserOptions: ParserOptions = {
	sourceType: 'module',
	strictMode: false,
	tokens: true,
	plugins: [
		'decorators-legacy',
		'doExpressions',
		'objectRestSpread',
		'classProperties',
		'classPrivateProperties',
		'classPrivateMethods',
		'exportDefaultFrom',
		'exportNamespaceFrom',
		'asyncGenerators',
		'functionBind',
		'functionSent',
		'dynamicImport',
		'numericSeparator',
		'optionalChaining',
		'importMeta',
		'bigInt',
		'optionalCatchBinding',
		'throwExpressions',
		'nullishCoalescingOperator'
	]
}

export default function buildParse(options: ParserOptions = {}): {
	parse: (src: string) => bt.File
} {
	options = {
		...babelParserOptions,
		...options,
		plugins: [...(babelParserOptions.plugins || []), ...(options.plugins || [])]
	}
	return {
		parse(src: string): bt.File {
			return parse(src, options) as bt.File
		}
	}
}
