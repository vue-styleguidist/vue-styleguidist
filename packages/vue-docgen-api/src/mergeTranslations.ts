import * as path from 'path'
import traverse from 'traverse'
import { ComponentDoc } from './Documentation'

export function getTranslationsObjectFromFile(fullFilePath: string, translation: string) {
	const transFileName = `${path.basename(fullFilePath).replace(/\.\w+$/, '')}.${translation}.js`
	return require(path.join(path.dirname(fullFilePath), transFileName))
}

export function mergeTranslationsAsObject(docs: ComponentDoc, transObjects: any[]): ComponentDoc {
	const transObject = transObjects.find((tr: { name: string }) => tr.name === docs.exportName)
	traverse(transObject).forEach(function(traversedItem) {
		if (!traversedItem.description) {
			return
		}
		// if the element is in array
		if (this.key && !isNaN(parseInt(this.key, 10)) && this.parent) {
			if (!traversedItem.name) return
			// find path of array
			const arrayPath = this.parent.path
			// get element in doc
			let i = 0
			let obj = docs
			while (i < arrayPath.length) {
				obj = obj[arrayPath[i++]]
				// if element does not exist quit
				if (!obj) return
			}
			const arrayOfProps = obj
			const prop = arrayOfProps.find((p: any) => p.name === traversedItem.name)
			// update its description with the one provided
			if (traversedItem.description && prop) {
				prop.description = traversedItem.description
			}
		} else {
			const arrayPath = this.path
			// get element in doc
			let i = 0
			let obj = docs
			while (i < arrayPath.length) {
				obj = obj[arrayPath[i++]]
				// if element does not exist quit
				if (!obj) return
			}
			// update its description with the one provided
			if (traversedItem.description) {
				obj.description = traversedItem.description
			}
		}
	})
	return docs
}

export default function mergeTranslations(
	docs: ComponentDoc,
	fullFilePath: string,
	translation: string
): ComponentDoc {
	const transObjects = getTranslationsObjectFromFile(fullFilePath, translation)
	return mergeTranslationsAsObject(docs, transObjects)
}
