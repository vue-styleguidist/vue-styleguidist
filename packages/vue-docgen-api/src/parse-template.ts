import * as pug from 'pug'
import { ASTElement, ASTNode, compile, SFCBlock } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import cacher from './utils/cacher'

export interface TemplateParserOptions {
	functional: boolean
	rootLeadingComment: string
}

export type Handler = (
	documentation: Documentation,
	templateAst: ASTElement,
	options: TemplateParserOptions
) => void

export default function parseTemplate(
	tpl: SFCBlock,
	documentation: Documentation,
	handlers: Handler[],
	filePath: string
) {
	if (tpl && tpl.content) {
		const template =
			tpl.attrs && tpl.attrs.lang === 'pug'
				? pug.render(tpl.content, { filename: filePath })
				: tpl.content
		const ast = cacher(() => compile(template, { comments: true, optimize: false }).ast, template)
		const rootLeadingCommentArray = /^<!--(.+)-->/.exec(template.trim())
		const rootLeadingComment =
			rootLeadingCommentArray && rootLeadingCommentArray.length > 1
				? rootLeadingCommentArray[1].trim()
				: ''
		const functional = !!tpl.attrs.functional
		if (functional) {
			documentation.set('functional', functional)
		}

		if (ast) {
			traverse(ast, documentation, handlers, {
				functional,
				rootLeadingComment
			})
		}
	}
}

export function traverse(
	templateAst: ASTElement,
	documentation: Documentation,
	handlers: Handler[],
	options: TemplateParserOptions
) {
	const traverseAstChildren = (templateAst: ASTElement) => {
		if (templateAst.children) {
			for (const childNode of templateAst.children) {
				if (isASTElement(childNode)) {
					traverse(childNode, documentation, handlers, options)
				}
			}
		}
	}

	if (templateAst.type === 1) {
		handlers.forEach(handler => {
			handler(documentation, templateAst, options)
		})
		if (templateAst.if && templateAst.ifConditions) {
			// for if statement iterate through the branches
			templateAst.ifConditions.forEach(({ block }) => {
				traverseAstChildren(block)
			})
		} else {
			traverseAstChildren(templateAst)
		}
	}
}

function isASTElement(node: ASTNode): node is ASTElement {
	return !!node && (node as ASTElement).children !== undefined
}
