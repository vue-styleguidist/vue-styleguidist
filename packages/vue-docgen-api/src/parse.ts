import * as fs from 'fs'
import * as path from 'path'
import { parseComponent, SFCDescriptor } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import parseScript, { Handler as ScriptHandler } from './parse-script'
import parseTemplate, { Handler as TemplateHandler } from './parse-template'
import scriptHandlers from './script-handlers'
import templateHandlers from './template-handlers'
import cacher from './utils/cacher'

const ERROR_EMPTY_DOCUMENT = 'The passed source is empty'

export { ScriptHandler, TemplateHandler }

export interface ParseOptions extends DocGenOptions {
	filePath: string
	/**
	 * In what language is the component written
	 * @default undefined - let the system decide
	 */
	lang?: 'ts' | 'js'
}

export interface DocGenOptions {
	/**
	 * Which exported variables should be looked at
	 * @default undefined - means treat all dependencies
	 */
	nameFilter?: string[]
	/**
	 * What alias should be replaced in requires and imports
	 */
	alias?: { [alias: string]: string }
	/**
	 * What directories should be searched when resolving modules
	 */
	modules?: string[]
	/**
	 * Handlers that will be added at the end of the script analysis
	 */
	addScriptHandlers?: ScriptHandler[]
	/**
	 * Handlers that will be added at the end of the template analysis
	 */
	addTemplateHandlers?: TemplateHandler[]
}

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseFile(documentation: Documentation, opt: ParseOptions) {
	const source = fs.readFileSync(opt.filePath, {
		encoding: 'utf-8'
	})
	return parseSource(documentation, source, opt)
}

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseSource(documentation: Documentation, source: string, opt: ParseOptions) {
	const singleFileComponent = /\.vue$/i.test(path.extname(opt.filePath))
	let parts: SFCDescriptor | null = null

	if (source === '') {
		throw new Error(ERROR_EMPTY_DOCUMENT)
	}

	if (singleFileComponent) {
		parts = cacher(() => parseComponent(source), source)
	}

	// get slots and props from template
	if (parts && parts.template) {
		const extTemplSrc: string =
			parts && parts.template && parts.template.attrs ? parts.template.attrs.src : ''
		const extTemplSource =
			extTemplSrc && extTemplSrc.length
				? fs.readFileSync(path.resolve(path.dirname(opt.filePath), extTemplSrc), {
						encoding: 'utf-8'
				  })
				: ''
		if (extTemplSource.length) {
			parts.template.content = extTemplSource
		}
		const addTemplateHandlers: TemplateHandler[] = opt.addTemplateHandlers || []
		parseTemplate(
			parts.template,
			documentation,
			[...templateHandlers, ...addTemplateHandlers],
			opt.filePath
		)
	}

	const extSrc: string = parts && parts.script && parts.script.attrs ? parts.script.attrs.src : ''
	const extSource =
		extSrc && extSrc.length
			? fs.readFileSync(path.resolve(path.dirname(opt.filePath), extSrc), {
					encoding: 'utf-8'
			  })
			: ''

	const scriptSource = extSource.length
		? extSource
		: parts
			? parts.script
				? parts.script.content
				: undefined
			: source
	if (scriptSource) {
		opt.lang =
			(parts && parts.script && parts.script.attrs && parts.script.attrs.lang === 'ts') ||
			/\.tsx?$/i.test(path.extname(opt.filePath)) ||
			/\.tsx?$/i.test(extSrc)
				? 'ts'
				: 'js'

		const addScriptHandlers: ScriptHandler[] = opt.addScriptHandlers || []
		if (scriptSource) {
			parseScript(scriptSource, documentation, [...scriptHandlers, ...addScriptHandlers], opt)
		}
	}

	if (!documentation.get('displayName')) {
		// a component should always have a display name
		documentation.set('displayName', path.basename(opt.filePath).replace(/\.\w+$/, ''))
	}
}
