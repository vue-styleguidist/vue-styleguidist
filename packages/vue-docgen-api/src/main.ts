import { ComponentDoc, Documentation } from './Documentation'
import { DocGenOptions, parseFile, ParseOptions, parseSource as parseSourceLocal } from './parse'

export { TemplateParserOptions } from './parse-template'
export { ScriptHandler, TemplateHandler } from './parse'
export { ComponentDoc, DocGenOptions, ParseOptions, Documentation }

/**
 * Parse the components at filePath and return props, public methods, events and slots
 * @param filePath absolute path of the parsed file
 * @param opts
 */
export function parse(
	filePath: string,
	opts?: DocGenOptions | { [alias: string]: string }
): ComponentDoc {
	return parsePrimitive((doc, options) => parseFile(doc, options), filePath, opts)
}

/**
 * Parse the `source` assuming that it is located at `filePath` and return props, public methods, events and slots
 * @param filePath absolute path of the parsed file
 * @param opts
 */
export function parseSource(
	source: string,
	filePath: string,
	opts?: DocGenOptions | { [alias: string]: string }
): ComponentDoc {
	return parsePrimitive((doc, options) => parseSourceLocal(doc, source, options), filePath, opts)
}

function isOptionsObject(opts: any): opts is DocGenOptions {
	return !!opts && !!opts.alias
}

function parsePrimitive(
	createDoc: (doc: Documentation, opts: ParseOptions) => void,
	filePath: string,
	opts?: DocGenOptions | { [alias: string]: string }
): ComponentDoc {
	const doc = new Documentation()
	const options: ParseOptions = isOptionsObject(opts)
		? { ...opts, filePath }
		: { filePath, alias: opts }
	createDoc(doc, options)
	return doc.toObject()
}
