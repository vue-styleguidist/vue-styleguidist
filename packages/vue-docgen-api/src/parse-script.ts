import { ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'
import { NodePath } from 'ast-types/lib/node-path'
import { parse } from 'recast'
import Map from 'ts-map'
import buildParser from './babel-parser'
import Documentation from './Documentation'
import { ParseOptions } from './parse'
import cacher from './utils/cacher'
import resolveExportedComponent from './utils/resolveExportedComponent'
import documentRequiredComponents from './utils/documentRequiredComponents'

const ERROR_MISSING_DEFINITION = 'No suitable component definition found'

export type Handler = (
	doc: Documentation,
	componentDefinition: NodePath,
	ast: bt.File,
	opt: ParseOptions
) => Promise<void>

export default async function parseScript(
	source: string,
	preHandlers: Handler[],
	handlers: Handler[],
	options: ParseOptions,
	documentation?: Documentation,
	forceSingleExport = false
): Promise<Documentation[] | undefined> {
	const plugins: ParserPlugin[] = options.lang === 'ts' ? ['typescript'] : ['flow']
	if (options.jsx) {
		plugins.push('jsx')
	}

	const ast = cacher(() => parse(source, { parser: buildParser({ plugins }) }), source)
	if (!ast) {
		throw new Error(`Unable to parse empty file "${options.filePath}"`)
	}

	const [componentDefinitions, ievSet] = resolveExportedComponent(ast)

	if (componentDefinitions.size === 0) {
		// if there is any immediately exported variable
		// resolve their documentations
		const docs = await documentRequiredComponents(documentation, ievSet, undefined, options)

		// if we do not find any compoents throw
		if (!docs.length) {
			throw new Error(`${ERROR_MISSING_DEFINITION} on "${options.filePath}"`)
		} else {
			return docs
		}
	}

	return executeHandlers(
		preHandlers,
		handlers,
		componentDefinitions,
		documentation,
		ast,
		options,
		forceSingleExport
	)
}

async function executeHandlers(
	preHandlers: Handler[],
	localHandlers: Handler[],
	componentDefinitions: Map<string, NodePath>,
	documentation: Documentation | undefined,
	ast: bt.File,
	opt: ParseOptions,
	forceSingleExport: boolean
): Promise<Documentation[] | undefined> {
	const compDefs = componentDefinitions
		.keys()
		.filter(name => name && (!opt.nameFilter || opt.nameFilter.indexOf(name) > -1))

	if (forceSingleExport && compDefs.length > 1) {
		throw Error(
			'vue-docgen-api: multiple exports in a component file are not handled by docgen.parse, Please use "docgen.parseMulti" instead'
		)
	}

	const docs = await Promise.all(
		compDefs.map(async name => {
			// If there are multiple exports and an initial documentation,
			// it means the doc is coming from an SFC template.
			// Only enrich the doc attached to the default export
			// NOTE: module.exports is normalized to default
			const doc =
				(compDefs.length > 1 && name !== 'default' ? undefined : documentation) ||
				new Documentation(opt.filePath)
			const compDef = componentDefinitions.get(name) as NodePath
			// execute all prehandlers in order
			await preHandlers.reduce(async (_, handler) => {
				await _
				return await handler(doc, compDef, ast, opt)
			}, Promise.resolve())
			await Promise.all(localHandlers.map(async handler => await handler(doc, compDef, ast, opt)))
			// end with setting of exportname
			// to avoid dependencies names bleeding on the main components,
			// do this step at the end of the function
			doc.set('exportName', name)
			return doc
		})
	)

	// default component first so in multiple exports in parse it is returned
	return docs.sort((a, b) =>
		a.get('exportName') === 'default' ? -1 : b.get('exportName') === 'default' ? 1 : 0
	)
}
