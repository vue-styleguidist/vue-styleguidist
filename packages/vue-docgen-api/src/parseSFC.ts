import { parse as parseComponent } from '@vue/compiler-sfc'
import * as path from 'path'
import { readFile } from 'fs'
import { promisify } from 'util'
import templateHandlers from './template-handlers'
import cacher from './utils/cacher'
import parseTemplate, { Handler as TemplateHandler } from './parse-template'
import Documentation from './Documentation'
import { ParseOptions } from './parse'
import parseScript from './parse-script'
import defaultScriptHandlers, { preHandlers } from './script-handlers'

const read = promisify(readFile)

export default async function parseSFC(
	initialDoc: Documentation | undefined,
	source: string,
	opt: ParseOptions
): Promise<Documentation[]> {
	let documentation = initialDoc

	// use padding so that errors are displayed at the correct line
	const { descriptor: parts } = cacher(() => parseComponent(source, { pad: 'line' }), source)

	// get slots and props from template
	if (parts.template) {
		const extTemplSrc = parts?.template?.attrs?.src

		const extTemplSource =
			extTemplSrc && typeof extTemplSrc === 'string' && extTemplSrc.length
				? await read(path.resolve(path.dirname(opt.filePath), extTemplSrc), {
						encoding: 'utf-8'
				  })
				: // if we don't have a content to the binding, use empty string
				  ''

		if (extTemplSource.length) {
			parts.template.content = extTemplSource
		}
		const addTemplateHandlers: TemplateHandler[] = opt.addTemplateHandlers || []

		documentation = initialDoc || new Documentation(opt.filePath)

		parseTemplate(parts.template, documentation, [...templateHandlers, ...addTemplateHandlers], opt)
	}

	const extSrc = (parts && parts.script && parts.script.attrs
		? parts.script.attrs.src
		: '') as string
	const extSource =
		extSrc && extSrc.length
			? await read(path.resolve(path.dirname(opt.filePath), extSrc), {
					encoding: 'utf-8'
			  })
			: ''

	const scriptSource = extSource.length
		? extSource
		: parts.script
		? parts.script.content
		: undefined
	opt.lang =
		(parts.script && parts.script.attrs && parts.script.attrs.lang === 'ts') ||
		/\.tsx?$/i.test(extSrc)
			? 'ts'
			: 'js'

	if (parts.customBlocks) {
		documentation = documentation || new Documentation(opt.filePath)

		const docsBlocks = parts.customBlocks
			.filter(block => block.type === 'docs' && block.content && block.content.length)
			.map(block => block.content.trim())

		if (docsBlocks.length) {
			documentation.setDocsBlocks(docsBlocks)
		}
	}

	const scriptHandlers = opt.scriptHandlers || [
		...defaultScriptHandlers,
		...(opt.addScriptHandlers || [])
	]

	const docs: Documentation[] = scriptSource
		? (await parseScript(
				scriptSource,
				opt.scriptPreHandlers || preHandlers,
				scriptHandlers,
				opt,
				documentation,
				initialDoc !== undefined
		  )) || []
		: // if there is only a template return the template's doc
		documentation
		? [documentation]
		: []

	if (documentation && !documentation.get('displayName')) {
		// a component should always have a display name
		// give a component a display name if we can
		const displayName = path.basename(opt.filePath).replace(/\.\w+$/, '')
		const dirName = path.basename(path.dirname(opt.filePath))
		documentation.set('displayName', displayName.toLowerCase() === 'index' ? dirName : displayName)
	}

	return docs
}
