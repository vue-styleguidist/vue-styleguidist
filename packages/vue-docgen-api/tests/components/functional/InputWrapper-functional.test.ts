import * as path from 'path'
import { ComponentDoc } from '../../../src/Documentation'
import { parse } from '../../../src/main'

const button = path.join(__dirname, './InputWrapper.vue')
let docButton: ComponentDoc

describe('tests button functional', () => {
	beforeEach(async () => {
		docButton = await parse(button)
	})

	it('should extract props from template if functional', () => {
		expect(docButton.props).toContainEqual(
			expect.objectContaining({ name: 'error', type: { name: 'boolean' } })
		)
		expect(docButton.props).toContainEqual(
			expect.objectContaining({ name: 'v-model', type: { name: 'string' } })
		)
		expect(docButton.props).toContainEqual(
			expect.objectContaining({ name: 'label', type: { name: 'string' } })
		)
	})

	it('should not return the value props as it is v-model', () => {
		expect(docButton.props).not.toMatchObject([{ name: 'value', type: { name: 'string' } }])
	})

	it('should match the snapshot', () => {
		expect(docButton).toMatchSnapshot()
	})
})
