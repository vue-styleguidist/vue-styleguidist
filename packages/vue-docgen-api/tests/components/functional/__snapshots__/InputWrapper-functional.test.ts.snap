// Vitest Snapshot v1

exports[`tests button functional > should match the snapshot 1`] = `
{
  "description": "Wrapper for input elements",
  "displayName": "InputWrapper",
  "docsBlocks": [
    "\`\` \`jsx <input-wrapper label=\\"Label\\" value=\\"–\\" id=\\"sdf\\"> … </input-wrapper>  \`\` \`",
  ],
  "events": undefined,
  "exportName": "default",
  "expose": undefined,
  "functional": true,
  "methods": undefined,
  "props": [
    {
      "defaultValue": {
        "func": false,
        "value": "false",
      },
      "description": "Should display error?",
      "name": "error",
      "type": {
        "name": "boolean",
      },
    },
    {
      "defaultValue": {
        "func": false,
        "value": "false",
      },
      "description": "Should the input be focused?",
      "name": "focused",
      "type": {
        "name": "boolean",
      },
    },
    {
      "defaultValue": {
        "func": false,
        "value": "''",
      },
      "description": "Prefix icon for input field",
      "name": "icon",
      "type": {
        "name": "string",
      },
    },
    {
      "defaultValue": {
        "func": false,
        "value": "false",
      },
      "description": "Should the input be disabled?",
      "name": "disabled",
      "type": {
        "name": "boolean",
      },
    },
    {
      "name": "value",
      "type": {
        "name": "undefined",
      },
    },
    {
      "defaultValue": {
        "func": false,
        "value": "false",
      },
      "description": "True if input label should always be \\"up there\\"",
      "name": "labelAlwaysFloating",
      "type": {
        "name": "boolean",
      },
    },
    {
      "description": "Prefix string for input field",
      "name": "prefix",
      "type": {
        "name": "string",
      },
    },
    {
      "description": "Input ID",
      "name": "id",
      "required": true,
      "type": {
        "name": "string|number",
      },
    },
    {
      "description": "String for label text",
      "name": "label",
      "required": true,
      "type": {
        "name": "string",
      },
    },
    {
      "defaultValue": {
        "func": false,
        "value": "''",
      },
      "description": "String with relevant error message",
      "name": "errorMessage",
      "type": {
        "name": "string",
      },
    },
    {
      "description": "Input value",
      "name": "v-model",
      "required": true,
      "tags": {
        "model": [
          {
            "description": true,
            "title": "model",
          },
        ],
      },
      "type": {
        "name": "string",
      },
    },
  ],
  "slots": [
    {
      "name": "default",
    },
  ],
  "sourceFiles": [
    /packages/vue-docgen-api/tests/components/functional/InputWrapper.vue,
  ],
  "tags": {},
}
`;
